### ISSUE TITLE
ResNet's and classify_one problems.

### ISSUE THREAD

**Issue by TimZaman** [NVIDIA/DIGITS#707]
I have been using ResNets lately with digits with great success.
So, the whole training/validation passes work great. Then i am trying one single image (or a lot), and then  the results are just very random, or puts one class 100.0% always.
I went investigating this further. 

Training and validation and then saving the epoch is done in `main.lua`.
The results from `main.lua` are saved and then later loaded in `test.lua` during a single image test.
The thing that is saved are the weights.
If i would not save just the weights, but the entire model to a t7 file, and load that in and assign it as a model in `test.lua`, the ResNet works again, and gives sensible values. 
So what must be happening is that something the ResNet needs is not copied and it not only the 'weights' variable but something else. I tested for gradients, but that wasn't the thing either.
What else can it be? Bias? Probably not, right.
Or might it be that the weight data is loaded in wrong?

**N.B.:**
I also noticed that i can save the entire model as a .t7 file. It would be around 20x as big as the weights. Calling `model:clearState()` actually reduces the model to 2x as big as just the weights, and if i save this, i have no problem with the ResNet when i load this in through test.lua. However, the package 'torch7-nv' is old and does not include 'model:clearState()' yet, so one would have to build torch from source.

**Comment by gheinrich** [comment]
If you're using the model from [there](https://github.com/facebook/fb.resnet.torch/blob/master/models/resnet.lua) or something similar then you have batch norm layers in your model. Those have parameters that are inferred from the training set but not learnt through gradient descent and not part of the parameters returned by `model:getParameters()`.

I have a patch for this on #701. Waiting for feedback. With the patch we will save the full model after calling `clearState()` if available. Otherwise we will resort to saving only the weights and biases.

Note that the patch will make it harder to deploy a CuDNN-trained model on CPU though.

**Comment by TimZaman** [comment]
Oh, &$(! I have been doing double work. Hilarious to see that your changes are almost identical to mine. Oh well, ditching mine and pulling yours. Thanks again, Greg. Sorry about this report then!
_Closed_
